### ISSUE TITLE
JSON-like syntax for instantiating data types

### ISSUE THREAD

**Issue by SimonCockx** [REGnosys/rosetta-dsl#398]
There is currently no way to explicitly instantiate a data type. As an example, suppose we want to model an employee of a company.
```
type Employee: 
  name           string   (1..1) 
  salary         number   (1..1) 
  isSeniorMember boolean  (1..1) 
  mentor         Employee (0..1) 
```

Currently, the only way to make an instance of this data type is by calling a function that returns an `Employee`. Before we can instantiate an `Employee`, we therefore must define a constructor function, which is quite verbose.

We therefore intend to introduce an intuitive JSON-like syntax to instantiate a data type. For example, we could instantiate an `Employee` called Dwight Schrute who has a mentor called Michael Scott as follows.
```
Employee { 
  name: "Dwight Schrute",
  salary: 4800.00,
  isSeniorMember: False,
  mentor: Employee {
    name: "Michael Scott",
    isSeniorMember: True,
    salary: 6300.00,
    mentor: empty
  }
} 
```

For a complete version of this proposal, see [this wiki page](https://github.com/REGnosys/rosetta-dsl/wiki/JSON-Syntax-Proposal).

**Comment by SimonCockx** [comment]
This syntax might also be useful to construct so-called *record types* such as `date`, e.g.,
```
date {
  day: 1,
  month: 11,
  year: 2022
}
```

**Comment by minesh-s-patel** [comment]
Would be be able to use Rosetta Expressions within the JSON like syntax?

Below is an example of calculating the `isSeniorMember` and getting the `salary` from another func.

```
type Person: 
  name           string   (1..1) 
  age              number   (1..1) 

type Employee: 
  person       Person   (1..1) 
  salary         number   (1..1) 
  isSeniorMember boolean  (1..1) 
  mentor         Employee (0..1) 

...
...

alias dwight: Person { 
  name: "Dwight Schrute",
  age: 25
} 

alias michael: Person { 
  name: "Michael Scott",
  age: 68
} 

set employee:  Employee { 
                              person: dwight,
                              salary: GetSalaryForPerson(dwight),
                              isSeniorMember: dwight -> age > 67,
                              mentor: Employee {
                                person: michael,
                                isSeniorMember: michael -> age > 67,
                                salary: GetSalaryForPerson(michael),
                                mentor: empty
                              }
                            } 
```

**Comment by SimonCockx** [comment]
@minesh-s-patel Definitely! I will make that clearer on the wiki.

**Comment by vjuge** [comment]
Following the [proposal](https://github.com/REGnosys/rosetta-dsl/wiki/JSON-Syntax-Proposal) on the wiki, I just can approve.
I'll add that cdm is IMO lacking a schema definition. Rosetta can be considered a sort of, but it's not useable in the code, sadly.

I'm thinking of some sort of json schema.
And having that, we could eventually store it along with the data in some kind of  [Avro](https://avro.apache.org/) format.
-> would be usefull for data archiving / persistence on DLTs (see what I mean :)
-> could be usefull for cdm versions upgrades

**Comment by SimonCockx** [comment]
@vjuge Thank you for the feedback. Automatically generating a schema for Rosetta data types is definitely worth discussing. I'll get back to this (perhaps in a separate issue).

**Comment by vjuge** [comment]
@SimonCockx that's cool, maybe I could give a try ? I should have some room early October

let me know if by then you already started (or I'll see by myself thanks to the issue been assigned and started...)

**Comment by SimonCockx** [comment]
@vjuge That would be great! I would be more than happy to help with setting things up, once you have some time. Give me a shout when you start.

**Comment by wandrille-moussel** [comment]
Hi, we would really love to see this implemented soon !

**Comment by SimonCockx** [comment]
Hi @wandrille-moussel, good news then! This feature has been implemented and it will be available from DSL v8.5.0 and onwards.
